grpclb_remote_balancer.go

```diff
@@ -0,0 +1,270 @@
+			LoadBalanceRequestType: &lbpb.LoadBalanceRequest_ClientStats{
+				ClientStats: stats,
+			},
+		}); err != nil {
+			grpclog.Errorf("grpclb: failed to send load report: %v", err)
+			return
+		}
+	}
+}
+
+func (lb *lbBalancer) watchRemoteBalancer() {
+	for {
+		select {
+		case <-lb.doneCh:
+			return
+		default:
+		}
+
+		lbClient := &loadBalancerClient{cc: lb.ccRemoteLB}
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		stream, err := lbClient.BalanceLoad(ctx, FailFast(false))
+		if err != nil {
+			grpclog.Errorf("grpclb: failed to perform RPC to the remote balancer %v", err)
+			continue
+		}
+
+		// grpclb handshake on the stream.
+		initReq := &lbpb.LoadBalanceRequest{
+			LoadBalanceRequestType: &lbpb.LoadBalanceRequest_InitialRequest{
+				InitialRequest: &lbpb.InitialLoadBalanceRequest{
+					Name: lb.target,
+				},
+			},
+		}
+		if err := stream.Send(initReq); err != nil {
+			grpclog.Errorf("grpclb: failed to send init request: %v", err)
+			continue
+		}
+		reply, err := stream.Recv()
+		if err != nil {
+			grpclog.Errorf("grpclb: failed to recv init response: %v", err)
+			continue
+		}
+		initResp := reply.GetInitialResponse()
+		if initResp == nil {
+			grpclog.Errorf("grpclb: reply from remote balancer did not include initial response.")
+			continue
+		}
+		if initResp.LoadBalancerDelegate != "" {
+			grpclog.Fatalf("TODO: Delegation is not supported yet.")
+		}
+
+		// streamDone will be closed by the readServerList goroutine when
+		// there's an error. So the sendLoadReport goroutine won't block on
+		// time.Ticker forever.
+		streamDone := make(chan struct{})
+
+		var wg sync.WaitGroup
+		wg.Add(2)
+		go func() {
+			defer wg.Done()
+			lb.readServerList(stream, streamDone)
+		}()
+		go func() {
+			defer wg.Done()
+			if d := convertDuration(initResp.ClientStatsReportInterval); d > 0 {
+				lb.sendLoadReport(stream, d, streamDone)
+			}
+		}()
+		wg.Wait()
+	}
+}
+
+func (lb *lbBalancer) dialRemoteLB(remoteLBName string) {
+	var (
+		dopts []DialOption
+	)
+	if creds := lb.opt.DialCreds; creds != nil {
+		if err := creds.OverrideServerName(remoteLBName); err == nil {
+			dopts = append(dopts, WithTransportCredentials(creds))
+		} else {
+			grpclog.Warningf("grpclb: failed to override the server name in the credentials: %v, using Insecure", err)
+			dopts = append(dopts, WithInsecure())
+		}
+	} else {
+		dopts = append(dopts, WithInsecure())
+	}
+	if lb.opt.Dialer != nil {
+		// WithDialer takes a different type of function, so we instead use a special DialOption here.
+		dopts = append(dopts, func(o *dialOptions) { o.copts.Dialer = lb.opt.Dialer })
+	}
+	// Explicitly set pickfirst as the balancer.
+	dopts = append(dopts, WithBalancerBuilder(newPickfirstBuilder()))
+	// Dial using manualResolver.Scheme, which is a random scheme generated
+	// when init grpclb. The target name is not important.
+	cc, err := Dial(lb.manualResolver.Scheme()+":///grpclb.server", dopts...)
```

Comment:
I'd like to see if there's some other way to do this.  Registering a random string as our resolver seems a bit hacky.  Maybe we can make a new, internal dial implementation that doesn't need a scheme but has a resolver builder passed to it directly?

Or can we use NewSubConn maybe (but never return it from the picker)?

Reply:
Can we make the scheme "`grpclb_internal_<random>`" or something like that instead so it's not _just_ a garble of random characters that will confuse users?

Reply:
Added `withResolverBuilder` dial option.